onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,709 @@
+          graph.IsOutput(node->OutputDefs()[0])) {
+        continue;
+      }
+      if (IsCastTo(node, TensorProto::FLOAT16)) {
+        cast_fp16_siblings.push_back(node);
+      } else if (IsCastTo(node, TensorProto::FLOAT)) {
+        cast_fp32_siblings.push_back(node);
+      }
+    }
+    if (cast_fp16_siblings.size() > 1) {
+      modified = true;
+      FuseNodes(graph, output, cast_fp16_siblings, removed_nodes);
+      VLOGS(logger, 1) << "FusedSubgraphs: Fused Cast nodes : " << ConcatNames<std::vector<Node*>>(cast_fp16_siblings);
+    }
+    if (cast_fp32_siblings.size() > 1) {
+      modified = true;
+      FuseNodes(graph, output, cast_fp32_siblings, removed_nodes);
+      VLOGS(logger, 1) << "FusedSubgraphs: Fused Cast nodes : " << ConcatNames<std::vector<Node*>>(cast_fp32_siblings);
+    }
+  }
+  return modified;
+}
+
+// RemoveUnnecessaryCasts
+// Remove a cast if the input elem_type is same the required cast type.
+static bool RemoveUnnecessaryCasts(Graph& graph, Node* node,
+                                   std::deque<onnxruntime::NodeIndex>& removed_nodes,
+                                   const logging::Logger& logger) {
+  bool modified = false;
+  if (node->InputDefs().size() == 1) {
+    const NodeArg* node_arg = node->InputDefs()[0];
+    auto elem_type = node_arg->TypeAsProto()->tensor_type().elem_type();
+    TensorProto_DataType data_type = static_cast<TensorProto_DataType>(elem_type);
+    if (IsCastTo(node, data_type)) {
+      VLOGS(logger, 1) << "Removed unnecessary cast " << node->Name();
+      RemoveCastNodes(graph, {node}, removed_nodes);
+      modified = true;
+    }
+  }
+  return modified;
+}
+
+// PropagateFP32CastsFromInputsToOutputs
+// This non recursive fusion, checks whether the given node is fp16 safe op and
+// whether all floatingpoint inputs are cast to fp32
+// and propagates cast op to the floatingpoint outputs.
+static bool PropagateFP32CastsFromInputsToOutputs(Graph& graph, Node* node,
+                                                  std::deque<onnxruntime::NodeIndex>& removed_nodes,
+                                                  size_t level,
+                                                  const logging::Logger& logger) {
+  bool modified = false;
+  if (IsFP16Allow(node->OpType(), level)) {
+    bool has_float_inputs = false;
+    bool all_float_inputs_have_casts = true;
+    std::vector<Node*> casts;
+    std::unordered_set<NodeArg*> require_type_change;
+    for (NodeArg* input : node->MutableInputDefs()) {
+      if (!IsType(*input, TensorProto::FLOAT)) {
+        continue;
+      }
+      has_float_inputs = true;
+      Node* producer = graph.GetMutableProducerNode(input->Name());
+      if (producer && IsCastTo(producer, TensorProto::FLOAT)) {
+        casts.push_back(producer);
+        require_type_change.insert(input);
+        continue;
+      }
+      all_float_inputs_have_casts = false;
+      break;
+    }
```

Comment:
Since we are not considering the fp32 initializer case yet... we can use following util function

auto casts = graph_utils::FindParentsByType(*node, "Cast");
bool all_float_inputs_have_casts = (casts.size() == node->NumInputs()) and (All casts are to Float);

Reply:
all_float_inputs_have_casts is also a stronger condition than has_float_inputs. 
so we acctually don't need has_float_inputs. 

---
In reply to: [612133729](https://github.com/microsoft/onnxruntime/pull/7191#discussion_r612133729) [](ancestors = 612133729)

Reply:
Also add a TODO here, to handle the non-uniform input types. 

---
In reply to: [612134201](https://github.com/microsoft/onnxruntime/pull/7191#discussion_r612134201) [](ancestors = 612134201,612133729)

Reply:
Actually all_float_inputs_have_casts  does not imply that the node has any float inputs.

Reply:
Also we want to allow the node to have non-float inputs casted or not.

Reply:
Also we don't need casts.size() == node->NumInputs() because the node may have non-float inputs, such as int, ints, etc.